pkg/ebpf/probes/tc_probes.go

```diff
@@ -0,0 +1,198 @@
+	DetachAll() error
+	DisableAutoload() error
+}
+
+type tcProbes struct {
+	module   *bpf.Module
+	tcProbes map[*net.Interface][]tcProbe
+}
+
+func InitTc(module *bpf.Module) (TcProbes, error) {
+	return &tcProbes{
+		tcProbes: make(map[*net.Interface][]tcProbe),
+		module:   module,
+	}, nil
+}
+
+// addInterface adds all default tcProbes to a given Interface
+func (p *tcProbes) addInterface(netIface *net.Interface) error {
+
+	// loopback devices special case (ingress only usually)
+	if netIface.Flags&net.FlagLoopback == net.FlagLoopback {
+		if _, ok := p.tcProbes[netIface]; !ok {
+			p.tcProbes[netIface] = loTcProbes[:]
+		}
+
+		return nil
+	}
+
+	if _, ok := p.tcProbes[netIface]; !ok {
+		p.tcProbes[netIface] = allTcProbes[:]
+	}
+
+	return nil
+}
+
+// delInterface deletes all default tcProbes from a given Interface
+func (p *tcProbes) delInterface(netIface *net.Interface) error {
+
+	delete(p.tcProbes, netIface)
+
+	return nil
+}
+
+// AttachToInterface attaches all tcProbes to their eBPF progs and Interfaces
+func (p *tcProbes) AttachToInterface(netIface *net.Interface) error {
+
+	p.addInterface(netIface)
+
+	for _, tcProbe := range p.tcProbes[netIface] {
+		err := tcProbe.attach(p.module, netIface)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// DetachFromInterface detaches all tcProbes associated with an Interface
+func (p *tcProbes) DetachFromInterface(netIface *net.Interface) error {
+
+	for _, tcProbe := range p.tcProbes[netIface] {
+		err := tcProbe.detach()
+		if err != nil {
+			return err
+		}
+	}
+
+	p.delInterface(netIface)
+
+	return nil
+}
+
+// DetachAll detaches all existing tcProbes
+func (p *tcProbes) DetachAll() error {
+
+	for netIface := range p.tcProbes {
+		p.DetachFromInterface(netIface)
+	}
+
+	return nil
+}
+
+// DisableAutoload disables all tcProbes programs autoload
+func (p *tcProbes) DisableAutoload() error {
+
+	for _, tcProbe := range allTcProbes {
+		tcProbe.disableAutoload(p.module)
+	}
+
+	for _, tcProbe := range loTcProbes {
+		tcProbe.disableAutoload(p.module)
+	}
+
+	return nil
+}
+
+// tcProbe
+
+type tcProbe struct {
+	tcHook      *bpf.TcHook
+	attachPoint bpf.TcAttachPoint
+	programName string
+}
+
+// attach attaches a tcProbe to a given Interface
+func (p *tcProbe) attach(module *bpf.Module, netIface *net.Interface) error {
+	var err error
+
+	p.tcHook = module.TcHookInit()
+
+	err = p.tcHook.SetInterfaceByName(netIface.Name)
+	if err != nil {
+		return fmt.Errorf("failed to set tc hook interface: %v", err)
+	}
+
+	p.tcHook.SetAttachPoint(p.attachPoint)
+
+	err = p.tcHook.Create()
+	if err != nil {
+		if errno, ok := err.(syscall.Errno); ok && errno != syscall.EEXIST {
+			return fmt.Errorf("tc hook create: %v", err)
+		}
+	}
+
+	if module == nil {
+		fmt.Printf("eh null\n")
+	}
+
+	prog, _ := module.GetProgram(p.programName)
+	if prog == nil {
+		return fmt.Errorf("could not find tc program %s", p.programName)
+	}
+
+	var tcOpts bpf.TcOpts
```

Comment:
style nit: why not `tcOpts := bpf.TcOpts{ ProgFd: int(prog.GetFd()) }`?

Reply:
No special reason, I believe I just moved this code and did not touch it. I'll change it.